dspy/clients/base_lm.py

```diff
@@ -139,15 +136,29 @@ def copy(self, **kwargs):
     def inspect_history(self, n: int = 1):
         return pretty_print_history(self.history, n)
 
-    def update_global_history(self, entry):
-        if settings.disable_history:
+    def update_history(self, entry):
+        if settings.disable_history or settings.max_history_size == 0:
             return
 
-        if len(GLOBAL_HISTORY) >= MAX_HISTORY_SIZE:
+        # dspy.LM.history
+        if len(self.history) >= settings.max_history_size:
+            self.history.pop(0)
+
+        self.history.append(entry)
+
+        # Global LM history
+        if len(GLOBAL_HISTORY) >= settings.max_history_size:
```

Comment:
Does this work properly with the following code? iirc, we didn't make max history a setting configuration as it is not thread local variable

```python
for i in range(100):
  lm(**kwargs)
with dspy.context(max_history_size=5):
  lm(**kwargs)
```

Reply:
The earlier setting is a bit strange, basically max number of history is a constant. But given it a second thought, I agree that at least we shouldn't make this change into this PR for safety. Reverted this change.